### PR TITLE
Query: Enable materialzing collections of arbirtary shape in projection

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #15611
Will send out separate PR to enable tests and fix edge case bugs

This PR
- Split out collection initialization/population for include to separate method due to type constraints
- Swallow materialize collection navigation/ToList in projection. Rest of unknown methods, we will fall into client eval
- Changes SqlTranslator to try translate & lift subquery always, allowing it to translate non Queryable defined but translatable methods
- Make identifiers nullable always to identify null keys & no correlation

### CODE REVIEW

File: src/EFCore.Relational/Query/Pipeline/CollectionInitializingExperssion.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+using System;
+using System.Linq.Expressions;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Query.Expressions.Internal;
+using Microsoft.EntityFrameworkCore.Query.Internal;
+
+namespace Microsoft.EntityFrameworkCore.Relational.Query.Pipeline
+{
+    public class CollectionInitializingExperssion : Expression, IPrintable
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/Pipeline/CollectionInitializingExperssion.cs]: Typo in `Experssion`

File: src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs [hunk 2]
```diff
@@ -94,13 +94,32 @@ public override Expression Visit(Expression expression)
...
 
-                    if (expression is MethodCallExpression methodCallExpression
-                        && methodCallExpression.Method.Name == "MaterializeCollectionNavigation")
+                    if (expression is MethodCallExpression methodCallExpression)
                     {
-                        var result = _queryableMethodTranslatingExpressionVisitor.TranslateSubquery(methodCallExpression.Arguments[0]);
-                        var navigation = (INavigation)((ConstantExpression)methodCallExpression.Arguments[1]).Value;
-
-                        return _selectExpression.AddCollectionProjection(result, navigation);
+                        if (methodCallExpression.Method.Name == "MaterializeCollectionNavigation")
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs]: Magic string

File: src/EFCore/Query/Pipeline/CollectionShaperExpression.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Query.Expressions.Internal;
+using Microsoft.EntityFrameworkCore.Query.Internal;
+
+namespace Microsoft.EntityFrameworkCore.Query.Pipeline
+{
+    public class CollectionShaperExpression : Expression, IPrintable
```
Comment (AndriySvyryd) [inline src/EFCore/Query/Pipeline/CollectionShaperExpression.cs]: 👍

File: src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs]: Yes, till @maumar puts custom expression rather than methodcall.
